<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [2]:
# В целях обеспечения безопасности данных для подклчения, файл с этими данными 'connect.csv' не выкладывается в публичный репозиторий на GitHub
start_connection=pd.read_csv('connect.csv') 
display(start_connection)

name                    values
0    DBNAME               project_sql
1     USER               skillfactory
2  PASSWORD  cCkxxLVrDE8EbvjueeMedPKt
3     HOST             84.201.134.129
4     PORT                       5432

In [3]:
connection = psycopg2.connect(
   dbname=start_connection.iloc[0,1],
   user=start_connection.iloc[1,1],
   host=start_connection.iloc[3,1],
   password=start_connection.iloc[2,1],
   port=start_connection.iloc[4,1]
)


# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query_3_1 = f'''select  COUNT(id)
           from VACANCIES
           '''

In [5]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_1652\1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
query_3_2 = f'''select  COUNT(id)
           from EMPLOYERS
           '''

In [7]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_1652\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
query_3_3 = f'''select  COUNT(id)
           from areas
           '''

In [9]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_1652\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_3_4 = f'''select  COUNT(id)
           from INDUSTRIES
           '''

In [11]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_1652\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

# выводы по предварительному анализу данных
Предварительный анализ показал, что в базе project_sql содержится информация о 49197 вакансиях от 23501 работодателя из 1362 регионов (при этом регионы в базе данных это не субъекты РФ, а отдельные города РФ и иные страны мира), в 294 сферах деятельноси. Все таблицы базы даннных имеют разную длину, 

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
# текст запроса
query_4_1 = f'''select  count(v.id), a.name
           from VACANCIES v
           join areas a on a.id=v.area_id
           group by 2
           order by 1 desc
           limit 5
           '''

In [13]:
# результат запроса
df1 = pd.read_sql_query(query_4_1, connection)
df1

C:\Users\User\AppData\Local\Temp\ipykernel_1652\3687461476.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df1 = pd.read_sql_query(query_4_1, connection)


count             name
0   5333           Москва
1   2851  Санкт-Петербург
2   2112            Минск
3   2006      Новосибирск
4   1892           Алматы

In [14]:
import plotly.express as px
fig =px.bar(
    data_frame=df1, 
    x='name', 
    y="count",
    color='name',
    height=500, 
    width=800,
    title='Зависимость количества вакансий от региона')
fig.show()

fig.write_html('bar_1.html') 

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [15]:
# текст запроса
query_4_2=f'''select  count(v.id)
           from VACANCIES v
           where salary_from is not null or salary_to is not null'''

In [16]:
# результат запроса
df2 = pd.read_sql_query(query_4_2, connection)
df2

C:\Users\User\AppData\Local\Temp\ipykernel_1652\995399408.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [17]:
# текст запроса
query_4_3=f'''select  round(avg(v.salary_from)) salary_from_avg, round(avg(v.salary_to)) salary_to_avg
           from VACANCIES v'''

In [18]:
# результат запроса
df3 = pd.read_sql_query(query_4_3, connection)
df3

C:\Users\User\AppData\Local\Temp\ipykernel_1652\2850329352.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



salary_from_avg  salary_to_avg
0          71065.0       110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [19]:
# текст запроса
query_4_4=f'''select  count(*), 
                      schedule, 
                      employment
           from VACANCIES v
           group by 2,3
           order by 1 desc'''
           

In [20]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_1652\2022935265.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count          schedule           employment
0   35367       Полный день     Полная занятость
1    7802  Удаленная работа     Полная занятость
2    1593     Гибкий график     Полная занятость
3    1312  Удаленная работа  Частичная занятость
4     940    Сменный график     Полная занятость
5     569       Полный день           Стажировка
6     367    Вахтовый метод     Полная занятость
7     347       Полный день  Частичная занятость
8     312     Гибкий график  Частичная занятость
9     141       Полный день     Проектная работа
10    133  Удаленная работа     Проектная работа
11    116     Гибкий график           Стажировка
12    101    Сменный график  Частичная занятость
13     64  Удаленная работа           Стажировка
14     18     Гибкий график     Проектная работа
15     12    Сменный график           Стажировка
16      2    Вахтовый метод     Проектная работа
17      1    Сменный график     Проектная работа

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [21]:
# текст запроса
query_4_5=f'''select  experience, count(experience)
           from vacancies
           group by experience
           order by count(experience)'''

In [22]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_1652\2316260449.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

In [23]:
#data_5=df['count'].groupby('experience', as_index=False)
import plotly.express as px
fig =px.bar(
    data_frame=df, 
    x='experience', 
    y="count",
    color='experience',
    height=500, 
    width=800,
    title='Зависимость количества вакансий от требуемого опыта работы')
fig.show()

fig.write_html('bar_4.4.html') 

***

# выводы по детальному анализу вакансий
Из детального анализа видно, что количество вакансий распределено не равномерно, наибольшее количество вакансий в нашей базе данных сосредоточено в Москве. Практически половина вакансий без указания заработной платы. Стоит отметить, что от размера заработной платы часто зависит интерес соискателя к вакансии, поэтому это очень важный параметр.Средняя заработная плата находится в вилке от 71065.0 рублей до 110537.0 рублей.  Много это или мало? У каждого человека свои требования к размеру ожидаемой заработной платы, вместе с тем из детального анализа видно, что даже нижняя граница средней заработной платы значително превышает прожиточный минимум, установленный в РФ на 2023 год (14 375 рублей). Большая часть вакансий имеет традиционное сочетание  типа рабочего графика и типа трудоустройства, а именно: "Полный день -	Полная занятость". Так же на графике "Зависимость количества вакансий от требуемого опыта работы" видно, что самое распространенное требование к стажу работы от 1 года до 3 лет.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [24]:
# текст запроса
query_5_1=f'''select *
              from (
                    select ROW_NUMBER() OVER(order by count(*) desc) AS Row,
                           count(*) count_vacancie,
                           e.name
                    from vacancies v
                         LEFT JOIN employers e ON v.employer_id=e.id
                    group by e.name
                    order by count(*) desc
                    ) t
              where Row in (1,5)'''

In [25]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_1652\3834399874.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



row  count_vacancie           name
0    1            1933         Яндекс
1    5             331  Газпром нефть

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [26]:
# текст запроса
query_5_2=f'''select a.name,
                     count(e.name)
              from areas a
                   left join vacancies v on v.area_id=a.id
                   left join employers e on a.id=e.area 
              where v.id is null
              group by 1
              order by 2 desc
              limit 1'''

In [27]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_1652\2250345976.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  count
0  Россия    410

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [28]:
# текст запроса
query_5_3=f'''select 
                   count(distinct area_id) count_areas
              from vacancies  
              group by employer_id
              order by count_areas desc
              limit 1'''

In [29]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_1652\201667002.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count_areas
0          181

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [30]:
# текст запроса
query_5_4=f'''select 
                   count(*) count_employers
              from employers e 
                   left join employers_industries ei on e.id=ei.employer_id
              where ei.industry_id is null'''

In [31]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_1652\3493105328.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count_employers
0             8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [32]:
# текст запроса
query_5_5=f'''SELECT e.name
              FROM employers e 
                   left join employers_industries ei on e.id=ei.employer_id
              GROUP BY e.name
              HAVING count(*)=4
              ORDER BY e.name 
              OFFSET 2
              LIMIT 1'''

In [33]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_1652\388711819.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [34]:
# текст запроса
query_5_6=f'''select count(*)
              from employers e 
                   left join employers_industries ei on e.id=ei.employer_id
                   left join industries i on i.id=ei.industry_id
              where i.name ='Разработка программного обеспечения'
              '''

In [35]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_1652\821628714.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [36]:
# код для получения списка городов-милионников
from pprint import pprint
import requests 
from bs4 import BeautifulSoup 
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8' # Определяем адрес страницы
response = requests.get(url) 
page = BeautifulSoup(response.text, "lxml") 
item=page.find_all('div', class_='mw-parser-output')
items=[]
for n, i in enumerate(item, start=1):
    item_1=i.find('table', class_='standard sortable')
    for n, j in enumerate(item_1, start=1):
        item_2=j.find_all('a')
        for n, q in enumerate(item_2, start=1):
            items.append(q.text.strip())
list_city=tuple(items[4:18]+items[19:])
print(list_city)

('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Волгоград')


In [37]:
# текст запроса
query_5_7=f'''(select 
                     a.name,
                     count(*) cnt
               from vacancies v 
                    left join areas a on a.id=v.area_id  
                    left join employers e on e.id=v.employer_id
               where e.name ='Яндекс' and a.name in {list_city}
               group by a.name
               order by cnt)
               union all
               select 
                     'total', sum(cnt)
               from (select 
                           a.name,
                           count(*) cnt
               from vacancies v 
                    left join areas a on a.id=v.area_id
                    left join employers e on e.id=v.employer_id
               where e.name ='Яндекс' and a.name in {list_city}
               group by a.name
               order by cnt) t
               '''

In [38]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_1652\1965596796.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name    cnt
0              Омск   21.0
1         Челябинск   22.0
2        Красноярск   23.0
3         Волгоград   24.0
4    Ростов-на-Дону   25.0
5            Казань   25.0
6               Уфа   26.0
7            Самара   26.0
8         Краснодар   30.0
9           Воронеж   32.0
10      Новосибирск   35.0
11  Нижний Новгород   36.0
12     Екатеринбург   39.0
13  Санкт-Петербург   42.0
14           Москва   54.0
15            total  460.0

***

# выводы по анализу работодателей
Лидирует по количеству вакансий от Яндекс Москва (45 вакансий), и здесь количество вакансий в 2.6 раза больше чем в городе Омске, в котором присутствует минимальное количество вакансий (21 вакансия) и примерно в 2 раза больше чем в 56%  городов-миллионников, в которых представлены вакансии от Яндекс (в 9 из 16 городов из нашего списка). На первый взгляд причина очевидна, Москва - столица, город с наибольшим количеством жителей, и прочие факторы, отличающие этот город от остальных, но еще возможна и другая причина. Не исключено, что вакансии в отдаленных от Москвы городах реже высвобождаются - становятся открытыми для набора специалистов, попросу если работа хорошая и оплата конкурентная на уровне региона, то люди неохотно покидают хорошее место в поисках лучшего. Таковы особенности жизни "за пределами Садового кольца"))

В подтверждение своего мнения, что причина наибольшего количества вакансий в Москве по сравнению с удаленными от нее городами не только в наибольшем количестве населения позволю себе рассмотреть соотношение количества вакансий и жителей ну хотябы среди лидеров по вакансиям Москвы и Санкт-Петербурга. Так количество жителей Санкт_петербурга составляет всего 43% от количества жителей Москвы, в это же время количество вакансий в Санкт-Петерурге составляет 79% от количества вакансий в Москве.

В реальности факторов и причин различного количества вакансий в различных регионах  больше. Начиная от наличия офисов Яндекс, они находятся не во всех городах, которые в нашем списке (офисы есть в: Москве, Санкт-Петербурге, Екатеринбурге, Казани, Нижнем Новгороде, Новосибирске и Ростове-на-Дону), наличия специалистов с необходимым образованием и опытом, запросов по зарплате (зависит от региона, мы исследовали это в предыдущем проекте) и другие.  

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [39]:
# текст запроса
query_6_1=f'''select 
                    count(*) as "Количество вакансий"
              from vacancies 
              where name ilike '%data%' or name ilike '%данн%'
           '''

In [40]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_1652\366702258.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Количество вакансий
0                 1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [41]:
# текст запроса
query_6_2=f'''select 
                    count(*) as "Количество вакансий"
              from 
                  vacancies 
              where 
                   (name ilike '%data scientist%' or name ilike '%data science%' or name ilike '%исследователь данных%'or name ilike '%ML%' 
                    or name ilike '%machine learning%' or name ilike '%машинн%обучен%') and name not ilike '%HTML%' and (name ilike '%junior%' 
                    or experience ilike '%Нет опыта%' or employment ilike '%Стажировка%') 
           '''

In [42]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_1652\2267038862.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Количество вакансий
0                   51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [43]:
# текст запроса
query_6_3=f'''select 
                    count(*) as "Количество вакансий"
              from 
                  vacancies
              where 
                   (key_skills ilike '%SQL%' or key_skills ilike '%postgres%') and ((name ilike '%data scientist%' or name ilike '%data science%' 
                   or name ilike '%исследователь данных%'or name like '%ML%' or name ilike '%machine learning%' or name ilike '%машинн%обучен%') 
                   and name not ilike '%HTML%')
           '''

In [44]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_1652\3794299702.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Количество вакансий
0                  201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [45]:
# текст запроса
query_6_4=f'''select 
                    count(*) as "Количество вакансий"
              from  
                  vacancies
              where 
                  (key_skills ilike '%python%') and ((name ilike '%data scientist%' or name ilike '%data science%' or name ilike '%исследователь данных%'
                  or name like '%ML%' or name ilike '%machine learning%' or name ilike '%машинн%обучен%') and name not ilike '%HTML%')
           '''

In [46]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_1652\4264665179.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Количество вакансий
0                  351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [47]:
# текст запроса
query_6_5=f'''select 
                   round(avg(1+(length(key_skills)-length(Replace(key_skills, CHR(9), '')))), 2) as "Количество навыков"
              from 
                  vacancies
              where 
                   (name ilike '%data scientist%' or name ilike '%data science%' or name ilike '%исследователь данных%'or name like '%ML%' 
                   or name ilike '%machine learning%' or name ilike '%машинн%обучен%') and name not ilike '%HTML%'
           '''

In [48]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_1652\3654210517.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Количество навыков
0                6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [49]:
# текст запроса
query_6_6=f'''SELECT 
                   experience,
                   round(AVG(COALESCE(((salary_from + salary_to)/2), salary_from, salary_to, 0)))
              FROM
                   vacancies
              WHERE 
                   ((name ILIKE '%data scientist%') OR (name ILIKE '%data science%') OR (name ILIKE '%исследователь данных%') OR ((name LIKE '%ML%') 
                   AND (name NOT ILIKE '%HTML%')) OR (name ILIKE '%machine learning%') OR (name ILIKE '%машинн%обучен%')) AND ((salary_from IS NOT NULL) 
                   OR (salary_to IS NOT NULL))
              GROUP BY 
                    experience
            '''

In [50]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_1652\519194770.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



experience     round
0           Нет опыта   74643.0
1  От 1 года до 3 лет  139675.0
2       От 3 до 6 лет  243115.0

***

# выводы по предметному анализу
В рассматриваемых данных о вакансиях более полутора тысяч вакансий имеют отношение к данным среди них только 51 вакансия подходит для начинающего специалиста, что означает высокую конкуренцию среди начинающих специалистов в облати работы с данными. Кроме того, это говорит о том, что в данной сфере ценится опыт и широкий спектр знаний, которых может быть недостаточно у новичка, так в среднем в каждой вакансии, имеющей отношение к данным, указано более 6 ключеных навыков. Интересен и тот факт, что среди ключевых навыков знание Python встречается чаще чем знание SQL. 

# Общий вывод по проекту
### (дополнительные исследования)

In [51]:
# Дополнительные исследования
# Определить по каким профессиям чаще всего предлагается "Вахтовый метод" работы.
query_d_0=f'''SELECT 
                   v.name,
                   count(v.name)
              FROM 
                   vacancies v
                   join areas a on v.area_id=a.id
              WHERE 
                   v.schedule='Вахтовый метод'
              GROUP BY 
                   v.name
              ORDER BY 2 desc
              LIMIT 10
              '''
# Определить регионы в котором чаще всего предлагают работу вахтовым методом
query_d_1=f'''SELECT 
                    a.name,
                    count(a.name)
              FROM 
                  vacancies v
                  join areas a on v.area_id=a.id
              WHERE 
                   v.schedule='Вахтовый метод'
              GROUP BY 
                   a.name
              ORDER BY 2 desc
              limit 10
              '''

In [52]:
df = pd.read_sql_query(query_d_0, connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_1652\3000808468.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  count
0                   Сервисный специалист (гидравлик)     20
1                               Руководитель проекта     19
2                  Сервисный специалист (агрегатчик)     18
3                    Сервисный специалист (моторист)     15
4                            Системный администратор     14
5                                         Сверловщик     12
6                     Обрубщик литья на производство     10
7                        Расточник на завод в Калугу     10
8  Инженер технического обслуживания систем АПС, ...      8
9    Ведущий специалист по промышленной безопасности      8

In [53]:
# результат запроса
df = pd.read_sql_query(query_d_1, connection)
df

C:\Users\User\AppData\Local\Temp\ipykernel_1652\1011015965.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  count
0                       Красноярск     22
1                          Иркутск     21
2                              Уфа     17
3                      Новосибирск     16
4                     Екатеринбург     14
5                             Омск     13
6  Благовещенск (Амурская область)     12
7                           Тюмень     11
8                             Чита     11
9                      Владивосток      9

По версии редакция международного Бизнес-журнала Lindeal HeadHunter возглавляет топ-15 самых лучших и универсальных российских сайтов по подбору персонала. В пользу универсальности свидетельствуют данные из задания 3.4., в котором установлено, что на сайте представлены вакансии в 294 сферах деятельности. На основе рассмотренных данных можно составить представление о наиболее востребованных профессиях, в разрезе каждой профессии соискатель может проанализировать требуемые навыки и более объективно сформировать свои запросы относительно ожидаемой зарплаты, а также спланировать индивидуальный план развития в профессии. Для работадателей интересным может стать анализ предлагаемых зарплат, так как на этапе планирования проекта зарплата сотрудникам будет заложена как одна из статей расхода. Также полезным будет анализ ключевых навыков по вакансиям и для организаций действующих в сфере образования, так как большое количество запросов работодателеей по каким-либо ключевым навыкам - это сигнал образовательным учреждениям на разработку соответствующего образовательного курса. 

В ходе работы над проектом по анализу данных о вакансиях стало возможным убедиться, что получать данные с помощью SQL-запросов удобно, довольно быстро и даннный инструмент позволяет проанализировать качественные и количественные признаки данных. 

В качестве даполнительного исследования я взяла показатель "schedule", отражающий тип рабочего графика. Мне стало интересно рассмотреть график "Вахтовый метод". Этот тип графика довольно специфический, и часто предполагает, что сотрудник работает в течение 1-6 месяцев, а затем отправляется к постоянному месту жительства, а на его место заступает другой специалист на тот же срок и так далее. И в этой связи меня удивило вот что. Для многих привычно, что таким образом работают строители и люди других рабочих специальностей, однако из результата запроса видно, что на втором месте по количеству вакансий "Руководитель проекта"(19 вакансий)... Могу только предположить, что срок полного цикла запуска и реализации проекта равен вахтовой смене. При этом если посмотреть на города, лидирующие по количеству вакансий с подобным графиком, мы видим, что это города, расположенные на восток от столицы и ближайший Уфа, находится на расстоянии более 1000 км от Москвы. 

Более того, исследование данных с HeadHunter и подобных ресурсов на постоянной основе, может быть полезно для создания различных социальных проектов, как собственно проект "Цифровые профессии", который был создан, чтобы восполнить дефицит IT-кадров.   